doc/memos/rdm-draft-petry-riot-design-goals.md

```diff
@@ -0,0 +1,218 @@
+
+## Easy fit, power supply-free smart home devices
+
+This requires a node to last for a number of weeks without recharge, for example
+to continue operating if a user is on holiday. It should be able to send fairly
+continuous data over a widely supported wireless LAN protocol, in order to
+support commercial home gateways.
+
+## Communication daughterboards for low-power protocols
+
+This requires RIOT to support popular emerging low-power protocols, and to
+provide robust and stable support for commonly used microcontrollers and
+transcievers.
+
+## Shipment tracking
+
+This requires a node to last for a number of weeks without recharge, monitor a
+range of environmental conditions, and send data to a local gateway which will
+be connected to the vehicle's power source and handle GPS localization.
+
+## Wearables
+
+This requires a node to be extremely light and low power, and to send data to a
+public gateway over a very low bandwidth connection, for example LoRaWAN, or to
+the user's smartphone, for example over Bluetooth. The device would require
+charging perhaps every week.
+
+## Diagnostic collection and/or control of traditionally non-powered devices
+
+This requires a node to be able to be retrofitted to a non-powered device such
+as a personal transportation device or item of public infrastructure, to be able
+to harvest energy from that device or last for a number of weeks or months
+without charging, to perform sensing of environmental or mechanical parameters,
+and to send data to a user's phone or a public wide-area network.
+
+## Building heating, ventilation, and air conditioning (HVAC)
+
+This requires a node to be able to be retrofitted to an already existing
+building, which may involve wireless communication but will likely involve a
+provided power source to the node. It will involve sensing of environmental
+conditions such as temperature, air quality and composition, and be able to
+feed back data at a rate that allows control of those conditions.
+
+
+# 3. List of design philosophies
+
+Below are a list of our design philosophies, including, where appropriate,
+an indication of relative priorities.
+
+## Energy efficiency
+
+RIOT nodes should be able to survive for years without running out of power.
+This means that they need to be able to be extremely conservative with power,
+and harvest any available energy from the environment.
+
+## Real-time capabilities
+
+For non-control and low-frequency sensing applications, precise knowledge of the
+data capture time is not neccessary. Nevertheless, RIOT users may wish to use
+the operating system in control or high frequency data capture applications, and
+in such a case, harder real-time guarantees and more accurate timing should be
+achievable at the expense of power consumption and configuration difficulty.
+
+## Small memory footprint
+
+RIOT aims to support constrained devices in class 1 [1]. This means that
```

Comment:
class 0!

Reply:
Sure? On which device of RAM << 10 KiB runs any Internet-of-Things OS?

Reply:
RIOT happily serves CoAP via dual-stack IPv4/IPv6 using ~3kB RAM, with room for improvement. Not with gnrc, though.

Reply:
>  The common understanding is that class 0 devices are not citizens of the IoT.

The referenced RFC states that class 0 devices are probably so resource constrained that they are unable to communicate directly with the internet. So I think requiring full IPv4 or IPv6 with CoAP is not within the definition of a class 0 device anymore. If I read the rfc (maybe a bit too liberal) (a cluster of) sensor nodes communicating with a more capable proxy over a low level (wired) protocol could count as class 0 devices.

Reply:
Yep, but this is not an IoT citizen. Honestly, a device with a few dozen or hundred Bytes of RAM should not hold a neighbor cache.

Reply:
> The referenced RFC states that class 0 devices are probably so resource constrained that they are unable to communicate directly with the internet.

Is it class 0 if it is too constrained to communicate directly with the internet, or is it class 0 if it is too small to be acurately described by class 1?

Reply:
> Honestly, a device with a few dozen or hundred Bytes of RAM should not hold a neighbor cache.

And as a non-routing 6LN it does not need to according to RFC 6775 (but I guess this isn't what @kaspar030 is aiming at).

Reply:
> Yep, but this is not an IoT citizen. Honestly, a device with a few dozen or hundred Bytes of RAM should not hold a neighbor cache.

@tcschmidt Why this focus on "IoT citizen"? It is not mentioned in this paragraph or in this document, do you have a definition for this term for me? To me it sounds very constraining to limit the applications of RIOT to this "IoT citizen" term


> Is it class 0 if it is too constrained to communicate directly with the internet, or is it class 0 if it is too small to be acurately described by class 1?

[The RFC](https://tools.ietf.org/html/rfc7228) states:
> They are so severely constrained in memory and processing capabilities that most likely they will not have the resources required to communicate directly with the Internet in a secure manner (rare heroic, narrowly targeted implementation efforts notwithstanding).

Reply:
Class 0, C0 | << 10 KiB             | << 100 KiB    
... as the fine manual states.

Reply:
> And as a non-routing 6LN it does not need to according to RFC 6775 (but I guess this isn't what @kaspar030 is aiming at).

Actually, was starting to implement a very reduced 6LN-only IPv6 stack in my spare time. Sadly, didn't had the time yet to come back to it (or back @kaspar030's claims here). But in theory a 6LN-only stack would only need a default route, an auto-configured link-local and global address (which can be for maximum RAM-saving also inferred just from a 6Lo compression context + link-local address). Maybe also some additional state.

Reply:
Mhmm, I could think of activities a lot more interesting and even more relevant. A good, full-featured stack implementation should suffice if it is modular enough. Sending hard-coded packets with magically numbered MAC addresses is not inspiring for networking people ;)

Reply:
> Mhmm, I could think of activities a lot more interesting and even more relevant. A good, full-featured stack implementation should suffice if it is modular enough. Sending hard-coded packets with magically numbered MAC addresses is not inspiring for networking people ;)

That was not my intend with that implementation. A full-featured stack was not the intention of my experiments there. Just the minimal requirements of what RFC 6775 asks from a non-routing 6LN. Modularity sometimes makes assumptions of other modules being there and thus need some code around it. A somewhat monolithic stack tailored just to 6LN+CoAP-server functionality might just fit those constraints without hardcoding magic numbers ;-) (ideally they are vendor supplied in ROM anyways).

Reply:
(because of lacking client functionality this wouldn't be compatible to CoAP-RD of course)

Reply:
> @tcschmidt Why this focus on "IoT citizen"? It is not mentioned in this paragraph or in this document, do you have a definition for this term for me? To me it sounds very constraining to limit the applications of RIOT to this "IoT citizen" term

A lot of people (including me) consider a key value of RIOT that it is an *IoT* operating system. Full-featured Internet-working by GNRC is one of its most important capabilities. Of course, one could strip this all off, but then we are approaching a state of very limited functionality which many 'firmwares' can provide. Just reading a sensor and blinking an LED is the ideal use case of Arduino. RIOT's use cases are those that subsume challenges of distributed systems.

Reply:
> 
> 
> > Mhmm, I could think of activities a lot more interesting and even more relevant. A good, full-featured stack implementation should suffice if it is modular enough. Sending hard-coded packets with magically numbered MAC addresses is not inspiring for networking people ;)
> 
> That was not my intend with that implementation. A full-featured stack was not the intention of my experiments there. Just the minimal requirements of what RFC 6775 asks from a non-routing 6LN. Modularity sometimes makes assumptions of other modules being there and thus need some code around it. A somewhat monolithic stack tailored just to 6LN+CoAP-server functionality might just fit those constraints without hardcoding magic numbers ;-) (ideally they are vendor supplied in ROM anyways).

The challenge is probably to get the modularity right in the sense that little to no overhead or glue code is needed. I would rather focus on even further improving and optimizing GNRC than hacking "something small" at the side. Because: GNRC is a cool concept and a real invention ...

Reply:
> A lot of people (including me) consider a key value of RIOT that it is an IoT operating system. Full-featured Internet-working by GNRC is one of its most important capabilities. Of course, one could strip this all off, but then we are approaching a state of very limited functionality which many 'firmwares' can provide. Just reading a sensor and blinking an LED is the ideal use case of Arduino. RIOT's use cases are those that subsume challenges of distributed systems.

Then see it this way: If we can't run RIOT without networking anymore with a most basic setup than we most definitely have a problem with our modularity.

Reply:
> Then see it this way: If we can't run RIOT without networking anymore with a most basic setup than we most definitely have a problem with our modularity.

This is getting me wrong: I was not saying that we *cannot* run RIOT without networking. What I meant: The interesting use cases for RIOT are with networking ... and we are discussing a (non-exhaustive) enumeration of *striking* use cases here.

Reply:
> > there are just some things you can't optimize away in GNRC without yielding really ugly code.
> 
> ... and/or without changing its design fundamentals.

I'm really not sure about this absolutist statement: changing the layout of sub-modules can change a lot. However, if we want to speak IPv6, we need to hold a 1280 Byte packet in a buffer. And I guess, we cannot do this without a little more than 1 KB of RAM. And I'm even not sure whether the packet buffer cannot be optimized for this single packet case without sacrificing its multi-packet handling.

Reply:
> I'm really not sure about this absolutist statement: changing the layout of sub-modules can change a lot. 

Well, GNRC concept is based around pktsnips and netapi. The former requires *multiple allocations* per packet. The latter requires *passing multiple messages* (as of now, even between threads and thus incurring *multiple context switches per packet*).

Both are setting a lower bound on memory consumption, runtime efficiency and predictability. Removing those is certainly possible, but the result would conceptually be a different network stack, and would certainly require much more rework than *changing the layout of sub-modules".

> However, if we want to speak IPv6, we need to hold a 1280 Byte packet in a buffer. And I guess, we cannot do this without a little more than 1 KB of RAM. And I'm even not sure whether the packet buffer cannot be optimized for this single packet case without sacrificing its multi-packet handling.

This is what happens if standards are not interpreted with actual use in mind. A non-routing node with only one defined application (e.g., CoAP) does *not* need to support 1280b MTU in order to be standards compliant. It is *perfectly acceptable* to drop incoming packets larger than what the application would drop anyways.

Don't pin me at <=1KB of RAM, ~3 KB RAM is enough for an IoT citizen. That is still way below the ~10KB that RFC7228 states for class one. For a 6LN, even 1kB might be enough.

@tcschmidt I understand you're not interested in work resulting in lowering IoT minimal RAM usage. But are you opposed to having that aim in this RDM?

Reply:
> A non-routing node with only one defined application (e.g., CoAP) does not need to support 1280b MTU in order to be standards compliant. It is perfectly acceptable to drop incoming packets larger than what the application would drop anyways.

Interesting statement, I never heard this: Can you backup with a corresponding RFC?

Reply:
> @tcschmidt I understand you're not interested in work resulting in lowering IoT minimal RAM usage. But are you opposed to having that aim in this RDM?

This is getting me wrong: Of course we all follow the objective to optimize RAM (resources). However, I'm making the case against unrealistic design goals and non-interoperable implementations.

1. The common, reasonable understanding is that class 0 devices are not in the horizon of IoT nodes, and RIOT has never really addressed those.
2. Implementations that violate standards (and hence interoperability) for the sake of resource efficiency are IMO simply a fallback into the early 2000-years of WSN handicraft work. The objective of RIOT is to professionalize beyond this.

Reply:
> Interesting statement, I never heard this: Can you backup with a corresponding RFC?

[RFC 8200](https://tools.ietf.org/html/rfc8200#section-5):

>    IPv6 requires that every link in the Internet have an MTU of 1280
>    octets or greater.  This is known as the IPv6 minimum link MTU.  On
>    any link that cannot convey a 1280-octet packet in one piece, link-
>    specific fragmentation and reassembly must be provided at a layer
>    below IPv6.

The keyword is IMHO is "convey" here. It would be pretty odd and stickler-to-the-rules-ish to ask from a non-forwarding host (i.e. a host that does not *convey* packets) that it *has to* under all circumstances accept a package of 1280 byte size (or more if we consider fragmentation) if that's not at all within the parameters of this host's application (i.e. the application would in the end just throw it away anyway), just because the RFC says the IPv6 link needs to support 1280 byte.

Reply:
@miri64 this is simply saying that you need something like 6LowPAN - " link-specific fragmentation and reassembly *must* be provided at a layer below IPv6."
As far as I know, any IPv6 node must be able to at least receive 1280 Bytes packets. This actually has a pretty strong reasoning behind. Giving up the 1280 bounds IMO means leaving the IPv6 protocol standard - and do something for private use.

Reply:
Maybe I'm wrong, but it seems that this discussion is biased by the idea that IoT == IP based networking. Which it is not.
Think of LoRaWAN for example: `examples/lorawan` fits into 7kB of RAM (without optimization), so even if it's not <<10kB, it fits into class 0 devices. And LoRaWAN is considered an IoT protocol.

Reply:
> As far as I know, any IPv6 node must be able to at least receive 1280 Bytes packets. This actually has a pretty strong reasoning behind. Giving up the 1280 bounds IMO means leaving the IPv6 protocol standard - and do something for private use.

How is "receive" defined? If the node "receives" then silently drops the packet, is it an acceptable optimization to not write the packet in any buffer in the meantime?

Anyhow, if the interaction with a node that "violates" the standard is indistinguishable from a node that does not violate the standard, it couldn't matter less.

Reply:
> As far as I know, any IPv6 node must be able to at least receive 1280 Bytes packets. This actually has a pretty strong reasoning behind. Giving up the 1280 bounds IMO means leaving the IPv6 protocol standard - and do something for private use.

Mhh.. the that comes to my mind where the application is not the deciding factor are router advertisements. They can - even in a 6LoWPAN - get pretty big (prefix option, 6Lo context options, ABRO, etc), but are also link-specific. Every other case, where IPv6 packets can natively get bigger are either forwarding related (e.g. [per fragment extension headers](https://tools.ietf.org/html/rfc8200#section-4.5) or use-case related (e.g. IPsec extension headers).

Reply:
> How is "receive" defined? If the node "receives" then silently drops the packet, is it an acceptable optimization to not write the packet in any buffer in the meantime?
>
> Anyhow, if the interaction with a node that "violates" the standard is indistinguishable from a node that does not violate the standard, it couldn't matter less.

If the silent dropping of the packet results in the node not being able to join the network properly or decipher context information for other packets given in that packet (see my RA example above), it is distinguishable from an implementation not violating the standard.

Reply:
> 
> 
> > How is "receive" defined? If the node "receives" then silently drops the packet, is it an acceptable optimization to not write the packet in any buffer in the meantime?
> > Anyhow, if the interaction with a node that "violates" the standard is indistinguishable from a node that does not violate the standard, it couldn't matter less.
> 
> If the silent dropping of the packet results in the node not being able to join the network properly or decipher context information for other packets given in that packet (see my RA example above), it is distinguishable from an implementation not violating the standard.

"receive" means successfully receive and process. The very simple interoperability issue is: A remote node has no way of telling link-specific properties at its destination and we cannot rely on path MTU discovery.

Reply:
> Anyhow, if the interaction with a node that "violates" the standard is indistinguishable from a node that does not violate the standard, it couldn't matter less.

Well, it of course matters. The simple insight: these Deerings and Hindens actually have used a lot of brain to make communication between nodes a success. If we don't want successful communication, we can just cut out the entire stack - that saves resources.

Reply:
> "receive" means successfully receive and process. The very simple interoperability issue is: A remote node has no way of telling link-specific properties at its destination and we cannot rely on path MTU discovery.

I still think this is the wrong argument for a **non-forwarding** end-point of a path. Even if it would accept 1280 B sized packets it still can silently discard the packet for any other reason and thus it is, as @kaspar030 put it, indistinguishable from not accepting the 1280 B sized packet. A "too big" RA illustrates the problem better for me.

Reply:
> If the silent dropping of the packet results in the node not being able to join the network properly or decipher context information for other packets given in that packet (see my RA example above), it is distinguishable from an implementation not violating the standard.

I have to agree on that. So given we cannot stream-parse RA's, the lower bound for IPv6 memory usage goes up by 1KB. I then state that <4KB of memory is enough for an IoT citizen. That is still way below 10KB. Should we call it "class 0.5"?

Reply:
@aabadie not necessarily. The discussion is about what drives the community, or more precisely for which journey we need a community at all. In the context of Internet communication, it is about interoperability (liberally speaking, based on whatever spec). 

If you design an implementation that perfectly fits a specific use case, you can do this, but should the community rely or work on this specific use case? I don't think so. 

"A non-routing node with only one defined application (e.g., CoAP)" [@kaspar030] needs CoAP. Technically, CoAP packets allow for "payload data [which] extends from after the marker to the end of the UDP datagram, i.e., the Payload" [RFC 7252]

Anyhow, the RIOT community should not work on being interoperable only within their own community.

Reply:
> Mhh.. the only thing that comes to my mind where the application is not the deciding factor 

@miri64 applications also matter: a node does not know what exactly its correspondents will send. You are always free to only speak the way you want (or can handle - as long as you are within the standards), but your correspondent may neither know nor care about your deficits. This is the cores story of *interoperability* enabled by standards.

Reply:
> 
> 
> Maybe I'm wrong, but it seems that this discussion is biased by the idea that IoT == IP based networking. Which it is not.
> Think of LoRaWAN for example: `examples/lorawan` fits into 7kB of RAM (without optimization), so even if it's not <<10kB, it fits into class 0 devices. And LoRaWAN is considered an IoT protocol.

Strong statement: LoRA is a link-layer technology that is augmented by a propitiatory deployment setting called "LoRAWAN". The "Internet of Things" is based on Internet standards, an IoT node is expected to  speak IPv6.

Reply:
> @miri64 applications also matter: a node does not know what exactly its correspondents will send. You are always free to only speak the way you want (or can handle - as long as you are within the standards), but your correspondent may neither know nor care about your deficits. This is the cores story of interoperability enabled by standards.

Yes, but same goes for a sender that does not use the agreed-upon (size-optimized) data-format (i.e. standard) for the application.

Reply:
> The very simple interoperability issue is: A remote node has no way of telling link-specific properties at its destination and we cannot rely on path MTU discovery.

If the application protocol doesn't ever use "large" packets, the interoperability problem starts when Firefox tries to http-connect to the corresponding port. Would the remote speak the protocol as defined, no packets larger than the maximum packet size as specified by the protocol.

Reply:
> 
> > * The common, reasonable understanding is that class 0 devices are not in the horizon of IoT nodes, and RIOT has never really addressed those.
> 
> Yes, that was state of the art 2014. Are you saying that we should stay there, ignorant of the advances in implementations? Why not innovate and then update that RFC to reflect the actual lower bounds?
> 

Interesting statement. Maybe I missed the "advances in implementations". However, the statement "common, reasonable understanding is that class 0 devices are not in the horizon of IoT nodes" was still valid this spring. I can try to get an update in Bangkok about the latest disruptions.

Reply:
> Yes, but same goes for a sender that does not use the agreed-upon (e.g. size-optimized) data-format (i.e. standard) for the application.

I don't get this. Are you referring to a sender that does not speak 6LowPan on an 802.15.5 link?
Anyway, the entire story of Standards is that members who comply to them can interoperate. Standards are the *only* agreements here. If standards are violated, inter-operation may fail.

Reply:
> Are you referring to a sender that does not speak 6LowPan on an 802.15.4 link?

Example Standard:
Sender MUST send a UDP packet to port 8888 with the content "HELLO".
Receiver MUST reply to any UDP packet on port 8888 and content "HELLO" with the content "WORLD". It MUST silently drop all other UDP packets received on that port.

The RA issue aside, a node that *only* runs a server for Example Standard, what's the packet buffer size the network stack it needs?

Reply:
??? Unclear what this means: Of course you can send "Hello" on UDP port 8888, but a network stack that can only send "Hello" on UDP port 8888 is not a network stack, but an introductory programming exercise. It is simply of no use - except in the case of a handcrafted firmware that is deployed in a proprietary regime.

Reply:
> 
> 
> > "receive" means successfully receive and process. The very simple interoperability issue is: A remote node has no way of telling link-specific properties at its destination and we cannot rely on path MTU discovery.
> 
> I still think this is the wrong argument for a **non-forwarding** end-point of a path. Even if it would accept 1280 B sized packets it still can silently discard the packet for any other reason and thus it is, as @kaspar030 put it, indistinguishable from not accepting the 1280 B sized packet. A "too big" RA illustrates the problem better for me.

@miri64 this subject is completely independent of forwarding versus non-forwarding. If you cannot handle valid IPv6 packets, but must silently drop, then you miserably fail. Presenting a stack that cannot do this is IMO not even worth thinking about. It will at most work in a proprietary environment - and that do all the proprietary industrial protocols, as well. Why thinking about IPv6, when you can only speak to your kind? It would be a ridiculous waste to handle IPv6 addresses while you can just exchange bits with your immediate predefined correspondent!

Reply:
@tcschmidt there is more nuance between sending 1280B packets and sending bits...

Reply:
BTW does anyone know where the number 1280B came from. The RFC doesn't offer much explanation and the more I think about it, with all the arguments we had in this thread, it seems pretty arbitrary to me (yes it's *the* number given in the standard so there can't be any discussion about it in this context, but an answer to that question would be nice anyway).

Reply:
Don't get me wrong, I understand that the number needs to be somewhat largish for fragment-wise extension headers, so arbitrary might be the wrong word, but why isn't it e.g. 1024 (which would be the next binary round number)? Why isn't it 1500 (a number also mentioned in the RFC and used as maximum frame size for Ethernet e.g.)? Why 1024+256?

Reply:
@danpetry don't get lost here: The naked RIOT kernel consumes about 1.5 KiB memory ... so barely fits on a class 0 device. This entire discussion is  illusional. I can ask people in the IETF whether they ever heard of an IPv6-networked general purpose OS that fits on nodes with ~1KiB (not in my name, though ;) ). Even if so, it is certainly not RIOT. The class 0 target imo makes no sense  for RIOT.

Reply:
> Of course you can send "Hello" on UDP port 8888, but a network stack that can only send "Hello" on UDP port 8888 is not a network stack, but an introductory programming exercise. It is simply of no use - except in the case of a handcrafted firmware that is deployed in a proprietary regime.

I was describing a hypothetical protocol that sits on top of any network stack. The point is that if an implementation of that protocol would be the only application on an embedded IPv6 stack, that stack doesn't need 1280b buffers (RA issue aside). For something less hypothetical, replace the HELLO WORLD protocol by a simple coap server serving temperature values.

Reply:
> 
> 
> > Of course you can send "Hello" on UDP port 8888, but a network stack that can only send "Hello" on UDP port 8888 is not a network stack, but an introductory programming exercise. It is simply of no use - except in the case of a handcrafted firmware that is deployed in a proprietary regime.
> 
> I was describing a hypothetical protocol that sits on top of any network stack. The point is that if an implementation of that protocol would be the only application on an embedded IPv6 stack, that stack doesn't need 1280b buffers (RA issue aside). For something less hypothetical, replace the HELLO WORLD protocol by a simple coap server serving temperature values.

Yes I know. But what you are describing is your own private protocol that is implemented by the same person (OS) on both ends. This is of course possible, but called proprietary. This is not the Internet and this does not need IP. This can be done with RIOT (without any IP stack), but is very boooring (has been done for decades now without success). The interesting potential of RIOT is that it cooperates in a very heterogeneous IoT.

Reply:
> > This entire discussion is illusional. I can ask people in the IETF whether they ever heard of an IPv6-networked general purpose OS that fits on nodes with ~1KiB (not in my name, though ;) ). Even if so, it is certainly not RIOT. The class 0 target imo makes no sense for RIOT.
> 
> RIOT is not a general purpose OS. It is an embedded OS. Ask them if the think it is possible to have a 6LN in ~1KB.

RIOT is a general purpose IoT OS for constrained (embedded) nodes. RIOT didn't start with the aim to be sub-size and smaller than anything else on earth. And I don't think this would be a promising objective: The real 'smart dust' can do a handful of operations, at most, and these are better handcrafted - not only because of resources, but foremost because of complexity: My trivial code better remains trivial ... 

> @tcschmidt What is your message here? GNRC is _good enough_, let's optimize that a little more and keep "timely implementing coming standards"?
>
GNRC is pretty cool and it is a shame that we don't leverage its gifts: This repeated 'strategy' of keeping upper layer protocols 'stack-independent' has prevented us from having running stack compositions with relevant protocols like DTLS etc.. It's rather ridiculous.

Our measurements have shown that GNRC is doing quite well in comparison to competing stacks like lwip and uIP - also in terms of memory footprint. Of course it could be improved and optimized - we currently don't have the time for it. However, we are sitting on solid grounds with GNRC and have learned to love its recursive composability in the quite successful ICN context.
 
> I hate to break it to you, but there's a couple of other embedded IOT operating systems on the market. RIOT had an edge at some point, but that is long gone. 

Surprising remark: it is only very recently that RIOT started to deliver more accurately what it promised. We are still fighting the reputation flaws accumulated from the many shortcomings of the early days. So I would state the opposite: RIOT is on its way to deliver just right now.

Regarding companies with money: We are used to trading brain against money since years. It hasn't been too bad, though.

> We can now look at RFC from 2014 stating class 0 devices cannot do proper internet, make it our doctrine and keep on "timely implementing standards".

You're getting me wrong: First, I talked to the authors of this RFC about the same subject only this year and didn't learn about apparent reasons for a change (I wonder what that should be anyway). Second, I'm not making it a 'doctrine'. Instead, we are discussing a general design document for RIOT and this document just would not carry much weight if filled with unrealistic and counter-productive content. Demanding class-0 compliance would wipe out a huge code portion off the RIOT repo. On the contrary, no one objects modules to be extremely well optimized and ultra small.

Reply:
> Yes I know. But what you are describing is your own private protocol that is implemented by the same person (OS) on both ends. This is of course possible, but called proprietary. This is not the Internet and this does not need IP. This can be done with RIOT (without any IP stack), but is very boooring (has been done for decades now without success). The interesting potential of RIOT is that it cooperates in a very heterogeneous IoT.

No, I was trying to explain that there are perfectly in-standard (non-proprietary) deployments where e.g., a CoAP only sensor node never needs to transmit or receive "large" transport layer packets.

Reply:
> GNRC is pretty cool and it is a shame that we don't leverage its gifts: This repeated 'strategy' of keeping upper layer protocols 'stack-independent' has prevented us from having running stack compositions with relevant protocols like DTLS etc.. It's rather ridiculous.

Or, keeping upper layer protocols stack-independent allows us to use different network stacks where GNRC's goals don't fit.

It is all well and good that you're proud of GNRC, but it seems to me you're trying to base RIOT's design goals around GNRC's design goals. IMO that is completely unnecessary. GNRC has its place, but we should not limit RIOT's goals with it.

One of RIOT's greatest strengths is its modularity. Most RIOT papers start stating that it can run on <32bit devices starting with 1.5kB of RAM. That implies to me that one can do usable things on class 0 devices, with RIOT. Over time, that number goes will go down, and that just shows that RIOT keeps getting better optimized and modularized.

> Demanding class-0 compliance would wipe out a huge code portion off the RIOT repo. On the contrary, no one objects modules to be extremely well optimized and ultra small.

Noone demands class0 compliance for all code of RIOT. Just because one of RIOT's goals is to support class 0 devices doesn't mean that all of RIOT's functionality has to. That would be unrealistic or too limiting.

The smallest MCU's RIOT supports currently have 2kB of memory. Supporting them does have a maintanance cost, but it has a *huge* benefit: we're constantly forced to question every bit of resource we use. That shows the higher up the stack you look, and that is why it is important to keep them in the tree. At that point we can stand to it and mention it in our goals.

It seems to me that this class0/1 point is also quite related to the question whether RIOT should be considered a good choice for non-networked applications. If one doesn't think so, and with RIOT's current state of needing a class 1 device for networking, putting class 1 as a goal makes sense. Less so if there are applications running happily on 2kB devices and we think RIOT is a valid choice for them.

How about something like this:

```suggestion
RIOT aims to be usable on low-end devices currently starting at 2 KiB RAM, with increasing functionality on more powerful devices. For networked applications, RIOT currently needs a class 1 device. This means that usable RIOT applications in certain key use cases (remote sensing, for example) consume ~ 10 KiB of RAM and ~100 KiB of ROM, or less.
```

My main intended change is a) we're not dropping low-end devices from our goals, and b) the class 1 requirement gets the word "currently" attached, implying that we're not necessarily stuck there.

Reply:
> 
> 
> > Yes I know. But what you are describing is your own private protocol that is implemented by the same person (OS) on both ends. This is of course possible, but called proprietary. This is not the Internet and this does not need IP. This can be done with RIOT (without any IP stack), but is very boooring (has been done for decades now without success). The interesting potential of RIOT is that it cooperates in a very heterogeneous IoT.
> 
> No, I was trying to explain that there are perfectly in-standard (non-proprietary) deployments where e.g., a CoAP only sensor node never needs to transmit or receive "large" transport layer packets.

CoAP normally expects a full IPv6 layer of 1280 Bytes payload (I know there are some error recovery mechanisms for truncated packets). Of course correspondents could send shorter messages, but as an implementer we don't know what the correspondent (whom we did not implement) will do. So we need to comply to standards.

Reply:
> CoAP normally expects a full IPv6 layer of 1280 Bytes payload (I know there are some error recovery mechanisms for truncated packets). Of course correspondents could send shorter messages, but as an implementer we don't know what the correspondent (whom we did not implement) will do. So we need to comply to standards.

Don't mix up implementation and usage. Of course all our IPv6 network stacks will support 1280b packets on the implementation side. But most of them will probably support configuring a smaller buffer, should the application not need 1280b. (e.g., gnrc_minimal sets a packet buffer size <1280b).

> I know there are some error recovery mechanisms for truncated packets

And how do you get truncated packets? riiiight, there in the standard, transport layer receive buffers < 1280b and how to deal with them.

Reply:
> Don't mix up implementation and usage. 

??? At the time of implementation, even at the time of deployment, you don't know the exact usage. So nothing to mix up here.

> > I know there are some error recovery mechanisms for truncated packets
> 
> And how do you get truncated packets? riiiight, there in the standard, transport layer receive buffers < 1280b and how to deal with them.

Please let me refer to the fine manual: https://tools.ietf.org/html/rfc7252

Reply:
> Please let me refer to the fine manual: https://tools.ietf.org/html/rfc7252

Yup, 4.6, implementation note:

```
If an
      implementation is too constrained to allow for allocating the
      above-mentioned upper bound, it could apply the following
      implementation strategy for messages not using DTLS security:
      Implementations receiving a datagram into a buffer that is too
      small are usually able to determine if the trailing portion of a
      datagram was discarded and to retrieve the initial portion.  So,
      at least the CoAP header and options, if not all of the payload,
      are likely to fit within the buffer.  A server can thus fully
      interpret a request and return a 4.13 (Request Entity Too Large;
      see Section 5.9.2.9) Response Code if the payload was truncated.
```

If a request doesn't fit my buffers, returning 4.13 (Request Entity Too Large) is alright.

Stop here as this is off-topic?

Reply:
> Stop here as this is off-topic?

 I would agree

> RIOT aims to be usable on low-end devices currently starting at 2 KiB RAM, with increasing functionality on more powerful devices. For networked applications, RIOT currently needs a class 1 device. This means that usable RIOT applications in certain key use cases (remote sensing, for example) consume ~ 10 KiB of RAM and ~100 KiB of ROM, or less.

I would remove the first "currently" and rephrase the second sentence to "For networked applications, class 1 devices are primarily supported". Adding "currently" feels like deferring the decision. We can always revise later in another RDM

Reply:
If i can jump in -- this thread is hitting close to home now. :-) I agree that RIOT should be able to handle a 1280b packet. However, I definitely agree that not every scenario needs to handle such a large packet size. Setting up a deployment that assumes a small packet size is a legitimate approach within an LLN where 6LoWPAN/802.15.4 packets are 128 bytes, or BLE packets can be even smaller. One of the coolest things added to RIOT's CoAP support has been the block extension (thanks Kaspar and Koen), where the application payload can be as small as 16 bytes. 

At the same time, at my day job I'm trying to convince people that MQTT on a cellular mote with TCP is a bad idea. For that, it would be nice to leverage RIOT's TCP support to implement the CoAP over TCP RFC. So, I support a composable system that allows development of a solution at both ends of the spectrum. I guess that's like Linux's support from embedded systems up to supercomputers.

Reply:
Just to be clear -- the block extension allows composing or decomposing a larger payload into smaller ones to adapt to these packet size constraints. :-)

Reply:
> 
> 
> Just to be clear -- the block extension allows composing or decomposing a larger payload into smaller ones to adapt to these packet size constraints. :-)

Yes, but this is for efficiency. It does not ensure that my neighbor does not send larger packets or without block mode. Let's recall the Internet design principle: "Be liberal in what you expect" ;)